x86_energy_plugin.cpp

```diff
@@ -0,0 +1,682 @@
+/*
+ * Copyright (c) 2016, Technische Universität Dresden, Germany
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification, are permitted
+ * provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this list of conditions
+ *    and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions
+ *    and the following disclaimer in the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse
+ *    or promote products derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#define ENV_CONFIG_PREFIX "SCOREP_METRIC_X86_ENERGY_PLUGIN_"
+#include <scorep/plugin/plugin.hpp>
+#include <scorep/plugin/util/matcher.hpp>
+
+#include <ctime>
+
+#include <string>
+#include <vector>
+#include <map>
+#include <unistd.h>
+#include <iostream>
+#include <chrono>
+#include <system_error>
+
+/* libs for the thread */
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+
+#include <ratio>
+
+extern "C" {
+#include <climits>
+#include <x86_energy.h>
+}
+
+
+using namespace scorep::plugin::policy;
+
+using scorep::plugin::logging;
+
+// Must be system clock for real epoch!
+using local_clock = std::chrono::system_clock;
+
+template <typename T>
+static double chrono_to_millis(T duration)
+{
+    return std::chrono::duration_cast<std::chrono::duration<double, std::milli>>(duration).count();
+}
+
+template <typename TP>
+class scorep_time_converter
```

Comment:
There you go:

https://github.com/score-p/scorep_plugin_cxx_wrapper/commit/371148ce5ff3074ab7153a5218ba6a698095ba47